Replaces http wasm middleware with more flexible implementation (#2922)
dapr/docs: daprdocs/content/en/reference/components-reference/supported-middleware/middleware-wasm.md

execute WebAssembly Modules (Wasm), which are most often binaries with a `.wasm`
extension.

> The Wasm [HTTP middleware]({{< ref middleware.md >}}) allows you to manipulate

Reviewer: Moving to one line.

```suggestion
The Wasm [HTTP middleware]({{< ref middleware.md >}}) allows you to manipulate an incoming request or serve a response with custom logic compiled to a Wasm binary. In other words, you can extend Dapr using external files that are not pre-compiled into the `daprd` binary. Dapr embeds [wazero][https://wazero.io] to accomplish this without CGO.
```

> an incoming request or serve a response with custom logic compiled to a Wasm

Reviewer: ```suggestion
```

> binary. In other words, you can extend Dapr using external files that are not

Reviewer: ```suggestion
```

> pre-compiled into the `daprd` binary. Dapr embeds [wazero][https://wazero.io]

Reviewer: ```suggestion
```

> to accomplish this without CGO.

Reviewer: ```suggestion
```

metadata:
  name: wasm
spec:
  type: middleware.http.wasm
  version: v1
  metadata:
  - name: path
    value: "./router.wasm"
```

## Spec metadata fields

> Minimally, a user must specify a Wasm binary implements the [http-handler](https://http-wasm.io/http-handler/).

Reviewer: ```suggestion
Minimally, you must specify a Wasm binary that implements the [http-handler](https://http-wasm.io/http-handler/).
```

> How to compile this is described later.

Reviewer: ```suggestion
[Learn how to compile your Wasm below]({{< ref "middleware-wasm.md#generating-wasm" >}}).
```

  httpPipeline:
    handlers:
    - name: wasm
      type: middleware.http.wasm
```

> *Note*: WebAssembly middleware uses more resources than native middleware. This

Reviewer: Moving note onto one line and using alert shortcode

```suggestion
{{% alert title="Note" color="primary" %}}
Wasm uses more resources than native middleware. This results in a resource constraint faster than the same logic in native code. Production usage should [control max concurrency]({{< ref control-concurrency.md >}}).
{{% /alert %}}
```

> result in a resource constraint faster than the same logic in native code.

Reviewer: ```suggestion
```

> Production usage should [Control max concurrency]({{< ref control-concurrency.md >}}).

Reviewer: ```suggestion
```

### Generating Wasm

> This component lets you manipulate an incoming request or serve a response with

Reviewer: ```suggestion
With the Wasm component, you can manipulate an incoming request or serve a response with custom logic compiled using the [`http-handler`](https://http-wasm.io/http-handler/) Application Binary Interface (ABI). The `handle_request` function receives an incoming request and can manipulate it or serve a response as necessary.
```

> custom logic compiled using the [http-handler](https://http-wasm.io/http-handler/)

Reviewer: ```suggestion
```

> Application Binary Interface (ABI). The `handle_request` function receives an

Reviewer: ```suggestion
```

> incoming request and can manipulate it or serve a response as necessary.

Reviewer: ```suggestion
```

> To compile your Wasm, you must compile source using a http-handler compliant

Reviewer: ```suggestion
To generate your Wasm, compile source using an `http-handler` compliant guest SDK, such as [TinyGo](https://github.com/http-wasm/http-wasm-guest-tinygo).
```

> guest SDK such as [TinyGo](https://github.com/http-wasm/http-wasm-guest-tinygo).

Reviewer: ```suggestion
```

Here's an example in TinyGo:

```go
package main

import (
	"strings"

	"github.com/http-wasm/http-wasm-guest-tinygo/handler"
	"github.com/http-wasm/http-wasm-guest-tinygo/handler/api"
)

func main() {
	handler.HandleRequestFn = handleRequest
}

// handleRequest implements a simple HTTP router.
func handleRequest(req api.Request, resp api.Response) (next bool, reqCtx uint32) {
	// If the URI starts with /host, trim it and dispatch to the next handler.
	if uri := req.GetURI(); strings.HasPrefix(uri, "/host") {
		req.SetURI(uri[5:])
		next = true // proceed to the next handler on the host.
		return
	}

	// Serve a static response
	resp.Headers().Set("Content-Type", "text/plain")
	resp.Body().WriteString("hello")
	return // skip the next handler, as we wrote a response.
}
```

> If using TinyGo, compile as shown below and set the spec metadata field named

Reviewer: ```suggestion
If using TinyGo, compile using the following command. Set the spec metadata field named `path` to the location of the output (for example, `router.wasm`):
```

> "path" to the location of the output (ex "router.wasm"):

Reviewer: ```suggestion
```

  httpPipeline:
    handlers:
    - name: wasm
      type: middleware.http.wasm.httpwasm
```

### Generating Wasm

This component lets you manipulate an incoming request or serve a response with
custom logic compiled using the [http-wasm][1] Application Binary Interface
(ABI). The `handle_request` function receives an incoming request and can
manipulate it or serve a response as necessary.

To compile your Wasm, you must compile source using a http-handler compliant
guest SDK such as [TinyGo](https://github.com/http-wasm/http-wasm-guest-tinygo).

Here's an example in TinyGo:

```go
package main

import (
	"strings"

	"github.com/http-wasm/http-wasm-guest-tinygo/handler"
	"github.com/http-wasm/http-wasm-guest-tinygo/handler/api"
)

func main() {
	handler.HandleRequestFn = handleRequest
}

// handleRequest implements a simple HTTP router.
func handleRequest(req api.Request, resp api.Response) (next bool, reqCtx uint32) {
	// If the URI starts with /host, trim it and dispatch to the next handler.

> 	if uri := req.GetURI(); strings.HasPrefix(uri, "/host") {

Reviewer 2: Could you follow invoke API or use substring to check, so that this example can work as expected?

Author: This documentation applies to the new, not the old one https://github.com/dapr/components-contrib/pull/2239
I'm required to make a PR on docs before it is merged, that's why this is here now. The example works, just not until the new PR is merged..

Author: This documentation applies to the new, not the old one https://github.com/dapr/components-contrib/pull/2239
I'm required to make a PR on docs before it is merged, that's why this is here now. The example works, just not until the new PR is merged..

Reviewer 2: Sure.

Author: ok components-contrib is up to date now
